datastore/src/main/java/org/jboss/pnc/datastore/repositories/internal/AbstractRepository.java

```diff
@@ -110,4 +116,54 @@ public int count(Collection<Predicate<T>> andPredicates, Collection<Predicate<T>
                 .getContent();
     }
 
+    /**
+     * @see Repository#cascadeUpdates) for full docs
+     * 
+     * @param managedNonOwning current version of entity from DB (MUST be Hibernate managed (due to LAZY fetching))
+     * @param updatedNonOwning proposed version of entity from request
+     * @param collectionGetter getter with collection of owning side (f.e ProductVersion::getBuildConfigurations)
+     * @param owningSetter setter which updates the owning side (f.e BuildConfiguration::setProductVersion)
+     * @param filter
+     * @param <N>
+     */
+    public <N extends GenericEntity<ID>> void cascadeUpdates(
+            N managedNonOwning,
+            N updatedNonOwning,
+            Function<N, Collection<T>> collectionGetter,
+            BiConsumer<T, N> owningSetter,
+            java.util.function.Predicate<T>... filter) {
+        // we want only to compare IDs because the data from request can be partial (due to dto refs in Maps)
+        Comparator<T> idComparator = (t, t1) -> t.getId().equals(t1.getId()) ? 0 : 1;
```

Comment:
This will break the TreeSet as `if A != B  then A > B && A < B`. (or in best case will be as efficient as array list)

You would want to use `Map<ID, Entity>` I think.

Reply:
Yes, TreeSet will be inefficient, you are absolutely correct. Map with ID as a Key basically does exactly what I need. Thx for the suggestion.